core/src/main/scala/kafka/idempotence/ProducerIdMapping.scala

```diff
@@ -0,0 +1,333 @@
+ * appended to the partition.
+ *
+ * @param topicPartition
+ * @param config
+ * @param snapParentDir
+ */
+@nonthreadsafe
+class ProducerIdMapping(val topicPartition: TopicAndPartition,
+                        val config: LogConfig,
+                        val snapParentDir: File) extends Logging {
+  import ProducerIdMapping._
+
+  val snapDir: File = new File(snapParentDir, formatDirName(topicPartition))
+  snapDir.mkdir()
+
+  private val pidMap = mutable.Map[Long, PidEntry]()
+  private var lastMapOffset = -1L
+  private var lastSnapOffset = -1L
+
+  /**
+   * Returns the last offset of this map
+   */
+  def mapEndOffset = lastMapOffset
+
+  /**
+   * Load a snapshot of the id mapping or return empty maps
+   * in the case the snapshot doesn't exist (first time).
+   *
+   * @return
+   */
+  private def loadFromSnapshot(logEndOffset: Long) {
+    pidMap.clear()
+    lastSnapshotFile(logEndOffset) match {
+      case Some(file) =>
+        try {
+          loadSnapshot(file, pidMap)
+          lastSnapOffset = offsetFromFile(file)
+          lastMapOffset = lastSnapOffset
+        } catch {
+          case e: CorruptSnapshotException =>
+            // TODO: Delete the file and try to use the next snapshot
+            throw e
+        }
+
+      case None =>
+        lastSnapOffset = -1
+        lastMapOffset = -1
+        snapDir.mkdir()
+    }
+  }
+
+  def truncateTo(logEndOffset: Long) = {
+    truncateSnapshotFiles(logEndOffset)
+    loadFromSnapshot(logEndOffset)
+  }
+
+  /**
+   * Update the mapping to the given epoch and sequence number.
+   * @param pid
+   * @param seq
+   * @param epoch
+   * @param offset
+   */
+  def update(pid: Long, seq: Int, epoch: Short, offset: Long) {
+    if (pid > 0) {
+      pidMap.put(pid, PidEntry(seq, epoch, offset))
+      lastMapOffset = offset
+    }
+  }
+
+  /**
+   * Verify the epoch and next expected sequence number. Invoked prior to appending the entries to
+   * the log.
+   * @param pid
+   * @param seq
+   * @param epoch
+   */
+  def checkSeqAndEpoch(pid: Long, seq: Int, epoch: Short) {
+    if (pid > 0) {
+      pidMap.get(pid) match {
+        case None =>
+          if (seq != 0)
+            throw new InvalidSequenceNumberException(s"Invalid sequence number: $pid (id), $seq (seq. number)")
+
+        case Some(tuple) =>
+          // Zombie writer
+          if (tuple.epoch > epoch)
+            throw new ProducerFencedException(s"Invalid epoch (zombie writer): $epoch (request epoch), ${tuple.epoch}")
+          if (tuple.epoch < epoch) {
+            if (seq != 0)
+              throw new InvalidSequenceNumberException(s"Invalid sequence number for new epoch: $epoch (request epoch), $seq (seq. number)")
+          } else {
+            if (seq != tuple.seq + 1L)
+              throw new InvalidSequenceNumberException(s"Invalid sequence number: $pid (id), $seq (seq. number), ${tuple.seq} (expected seq. number)")
```

Comment:
Ack. I wondered about this too. I left the existing logic as it is for now, but we do need to handle duplicates.